tools/stats/upload_external_contrib_stats.py

```diff
@@ -0,0 +1,143 @@
+import argparse
+import datetime
+import json
+import os
+import urllib.parse
+from typing import Any, Callable, cast, Dict, List, Optional
+from urllib.error import HTTPError
+from urllib.request import Request, urlopen
+
+from tools.stats.upload_stats_lib import upload_to_s3
+
+FILTER_OUT_USERS = set(["pytorchmergebot", "facebook-github-bot", "pytorch-bot[bot]"])
+
+
+def _fetch_url(
+    url: str,
+    *,
+    headers: Optional[Dict[str, str]] = None,
+    data: Optional[Dict[str, Any]] = None,
+    method: Optional[str] = None,
+    reader: Callable[[Any], Any] = lambda x: x.read(),
+) -> Any:
+    if headers is None:
+        headers = {}
+    token = os.environ.get("GITHUB_TOKEN")
+    if token is not None and url.startswith("https://api.github.com/"):
+        headers["Authorization"] = f"token {token}"
+    data_ = json.dumps(data).encode() if data is not None else None
+    try:
+        with urlopen(Request(url, headers=headers, data=data_, method=method)) as conn:
+            return reader(conn)
+    except HTTPError as err:
+        print(err.reason)
+        print(err.headers)
+        if err.code == 403 and all(
+            key in err.headers for key in ["X-RateLimit-Limit", "X-RateLimit-Used"]
+        ):
+            print(
+                f"Rate limit exceeded: {err.headers['X-RateLimit-Used']}/{err.headers['X-RateLimit-Limit']}"
+            )
+        raise
+
+
+def fetch_json(
+    url: str,
+    params: Optional[Dict[str, Any]] = None,
+    data: Optional[Dict[str, Any]] = None,
+) -> List[Dict[str, Any]]:
+    headers = {"Accept": "application/vnd.github.v3+json"}
+    if params is not None and len(params) > 0:
+        url += "?" + "&".join(
+            f"{name}={urllib.parse.quote(str(val))}" for name, val in params.items()
+        )
+    return cast(
+        List[Dict[str, Any]],
+        _fetch_url(url, headers=headers, data=data, reader=json.load),
+    )
+
+
+def get_external_pr_data(
+    start_date: datetime.date, end_date: datetime.date, period_length: int = 1
+) -> List[Dict[str, Any]]:
+    pr_info = []
+    period_begin_date = start_date
+
+    pr_count = 0
+    users = set([Dict[str, Any]])
+    while period_begin_date < end_date:
```

Comment:
Just a note: if querying GitHub API manually gets in the way, you could consider using https://github.com/khornberg/octokit.py to streamline the API call